src/scriptrag/api/index.py

```diff
@@ -0,0 +1,413 @@
+            force: Force re-indexing of all scripts
+            dry_run: Preview changes without applying them
+            batch_size: Number of scripts to process in each batch
+            progress_callback: Optional callback for progress updates
+
+        Returns:
+            IndexOperationResult with details of the operation
+        """
+        result = IndexOperationResult()
+
+        try:
+            # Check database exists
+            if not self.db_ops.check_database_exists():
+                error_msg = (
+                    "Database not initialized. Please run 'scriptrag init' first."
+                )
+                logger.error(error_msg)
+                result.errors.append(error_msg)
+                return result
+
+            # Step 1: Find all Fountain files with metadata
+            if progress_callback:
+                progress_callback(0.1, "Discovering Fountain files...")
+
+            scripts = await self._discover_scripts(path, recursive)
+
+            if not scripts:
+                logger.info("No Fountain files found")
+                return result
+
+            logger.info(f"Found {len(scripts)} Fountain files")
+
+            # Step 2: Filter scripts that need indexing
+            if not force:
+                scripts_to_index = await self._filter_scripts_for_indexing(scripts)
+            else:
+                scripts_to_index = scripts
+
+            if not scripts_to_index:
+                logger.info("No scripts need indexing")
+                return result
+
+            logger.info(f"Processing {len(scripts_to_index)} scripts")
+
+            # Step 3: Process scripts in batches
+            total = len(scripts_to_index)
+            for i in range(0, total, batch_size):
+                batch = scripts_to_index[i : min(i + batch_size, total)]
+                batch_num = (i // batch_size) + 1
+                total_batches = (total + batch_size - 1) // batch_size
+
+                if progress_callback:
+                    progress = 0.1 + (0.8 * i / total)
+                    progress_callback(
+                        progress, f"Processing batch {batch_num}/{total_batches}..."
+                    )
+
+                batch_results = await self._process_scripts_batch(batch, force, dry_run)
+                result.scripts.extend(batch_results)
+
+                # Collect errors from batch
+                for script_result in batch_results:
+                    if script_result.error:
+                        result.errors.append(
+                            f"{script_result.path}: {script_result.error}"
+                        )
+
+            if progress_callback:
+                progress_callback(1.0, "Indexing complete")
+
+        except Exception as e:
+            logger.error(f"Index operation failed: {e!s}")
+            result.errors.append(f"Index operation failed: {e!s}")
+
+        return result
+
+    async def _discover_scripts(
```

Comment:
This seems like a conceptual duplication of the ScriptLister functionality. Re-use, don't re-implement.